Minimize Bug Report Impact to Target

qskwood opened the pull request:
The bug report had the potential to install a handful of tools if not detected on the machine, including the following:
* ibstat (part of infiniband-diags)
* ipmitool (part of ipmitool)
* iostat (part of sysstat)
* lshw (part of lshw)
* sensors (part of lm-sensors)
* smartctl (part of smartmontools)

In addition to the comments raised by Josh. in https://github.com/lambdal-support/lambda-public-tools/pull/2#issuecomment-2332194673 about not installing software on instances, some of these tools are not always useful. For example, a cloud instance won't benefit from ipmitool or smartmontools because they are virtualized.

This PR establishes a few new behaviors:
1. it will, by default, not install any tools (setting the environment variable SKIP_TOOLS to 0 will override)
2. it will not install to a detected cloud instance any tools which would provide no benefit there
3. it functionalizes tool installation and the above checks instead of repeating code
4. it will gracefully (read: silently) handle tools that cannot be run where they may not be installed

I have also removed the install check for `infiniband-diags`, since to my knowledge any machine without `rmda-core` can't use IB and the latter package depends on the former.  I'm open to correction on this. I lastly modified the `nvidia-smi` parts to gracefully deal with `nvidia-smi` not being present. Even though they are not being install checked by the script, CPU nodes for example might not have it.

I did not remove it, but the install check for `lshw` is probably unnecessary. Both it and `dmidecode` are dependencies of `ubuntu-standard` and I suspect it will always be available. This might be something for future review. There are other enhancements I'd like to make to improve maintainability and readability that I chose not to include in this PR to keep it focused on one subject.

qskwood committed:
feat: only install packages useful for the target machine

Previously, the bug report script installed a handful of packages regardless of whether they were
likely to be useful on the target machine. This adds a check for whether a machine is a VM then,
if so, does not install packages that are unlikely to be useful on a VM.

qskwood committed:
feat: skip tool installation by default

Previously tools were installed by default, except for those unlikely to be useful on a VM and
when a VM is detected. This modifies the behavior so tools will only be installed if the
environment variable SKIP_TOOLS=0 is present.

qskwood committed:
refactor: adjust logic from 0=true to 1=true for VM check

Previously this logic was harder to read than necessary. This fixes that.

qskwood committed:
feat: expand tool installation checks for VMs

Previously the script assumed tools weren't useful for VMs. The script now checks both whether a
tool is useful for a VM and whether the script is targeting a VM before skipping installation.

qskwood committed:
feat: clean up stdout and document rationale

Previously the script would output errors about tools that were unavailable. Now the script doesn't
output errors for tools that may not have been installed by the script. The script also contains
documentation about how to silence these errors and a suggestion to note the tool was run but did
not produce any output, and why.

qskwood committed:
fix: remove infiniband-diags install check

rmda-core depends on infiniband-diags, so I don't believe we would ever need to install this. A
machine without rdma-core can't use IB, AFAICT, so we wouldn't care about IB information.

qskwood committed:
feat: gracefully handle the absence of nvidia-smi

Previously the script would produce errors if nvidia-smi wasn't avaialable, e.g., on a CPU node.
Now the script will direct nvidia-smi stderr to the same output file as stdout, and if the output
file is empty, will note in the ouptut file that nvidia-smi may not be available on the machine.

jashoakley reviewed (dismissed):
Tested with OD instance. Bug report ran successfully – verified after that no unnecessary packages were installed.

qskwood committed:
Merge branch 'main' into bug-report-min

qskwood committed:
chore: adapt merged commit to new behavior

A newly merged commit that uses nvidia-smi didn't redirect stderr or note in the output file why it
may be empty. This commit adds the new behavior to this older commit.

qskwood commented:
A new commit had an nvidia-smi command that needed the same stderr redirection. I've added this change.

jashoakley reviewed (approved):
LGTM
